### ISSUE TITLE
How to pretrain a imagenet network

### ISSUE THREAD

**Issue by SeaOfOcean** [sanghoon/pva-faster-rcnn#9]
Hi, I read from paper that PVANET is pre-trained with ILSVRC2012, which is an image classification task. How do you pre-train without bbox annotations?

**Comment by SeaOfOcean** [comment]
As mentioned in issue #1 , there is a shapre mismatch in imagenet pretrained model, can change the name of 'fc6' layer so that the params are re-initialized. If so, will it hurts the finetune effectiveness? By the way, what's the training dataset for full model? @sanghoon

**Comment by zhangjiangqige** [comment]
I have another question related to the pretraining. @sanghoon 
I'm using mxnet, and mxnet does not split bn into "bn+scale" as in caffe. I think bn in mxnet should be placed after the concatenation of conv, and before relu, is this right? This structure seems more similar to the crelu structure in your paper(figure 1) than the structure in the prototexts. Moreover, in my experiments, the network don't converge if bn are placed before concatenation..

**Comment by sanghoon** [comment]
@SeaOfOcean 
If you start fine-tuning with 're-initialized' FC layers,
it definitely hurts the resulted mAP. (1~2% or more?)
The training dataset for pre-training was the ILSVRC2012 training set.

@neodooth 
Please note that we haven't batch-normalized while fine-tuning.
The BN layers were effective only in the pre-training (ImageNet) phase.
From my experience,
putting BN layers before concatenation led to almost the same results.

**Comment by dopplesoldner** [comment]
I managed to train a model with weights from  Imagenet full using the prototxt in https://github.com/sanghoon/pva-faster-rcnn/commit/4799b2889fecdccd823559a7f0300bc68d69aefd

Note that I am using a custom dataset with 10 classes.

However, I have a shape mismatch problem at test time - 

```
I1012 19:03:05.922343 31279 net.cpp:283] Network initialization done.
I1012 19:03:06.218982 31279 net.cpp:761] Ignoring source layer input-data
I1012 19:03:06.219004 31279 net.cpp:761] Ignoring source layer data_input-data_0_split
I1012 19:03:06.219007 31279 net.cpp:761] Ignoring source layer im_info_input-data_1_split
I1012 19:03:06.219010 31279 net.cpp:761] Ignoring source layer gt_boxes_input-data_2_split
I1012 19:03:06.222738 31279 net.cpp:761] Ignoring source layer rpn_cls_score_rpn_cls_score_0_split
I1012 19:03:06.222795 31279 net.cpp:761] Ignoring source layer rpn_bbox_pred_rpn_bbox_pred_0_split
I1012 19:03:06.222798 31279 net.cpp:761] Ignoring source layer rpn_cls_score_reshape_rpn_cls_score_reshape_0_split
I1012 19:03:06.222800 31279 net.cpp:761] Ignoring source layer rpn-data
I1012 19:03:06.222813 31279 net.cpp:761] Ignoring source layer rpn_loss_cls
I1012 19:03:06.222816 31279 net.cpp:761] Ignoring source layer rpn_loss_bbox
I1012 19:03:06.222821 31279 net.cpp:761] Ignoring source layer mute_rpn_scores
I1012 19:03:06.222825 31279 net.cpp:761] Ignoring source layer roi-data
I1012 19:03:06.271338 31279 net.cpp:761] Ignoring source layer loss_cls
I1012 19:03:06.271358 31279 net.cpp:761] Ignoring source layer loss_bbox
F1012 19:03:06.354282 31279 concat_layer.cpp:42] Check failed: top_shape[j] == bottom[i]->shape(j) (38 vs. 37) All inputs must have the same shape, except at concat_axis.
*** Check failure stack trace: ***
./experiments/scripts/faster_rcnn_end2end.sh: line 85: 31279 Aborted                 (core dumped) ./tools/test_net.py --gpu ${GPU_ID} --def models/${PT_DIR}/${NET}/faster_rcnn_end2end/test.prototxt --net ${NET_FINAL} --imdb ${TEST_IMDB} --cfg experiments/cfgs/faster_rcnn_end2end.yml ${EXTRA_ARGS}
```

Any ideas on this? 
Thanks

**Comment by gjy3035** [comment]
@sanghoon  I want to train a pre-train pvanet model on ImageNet for classification task, but i do not have prototxt file. Do you have provide the prototxt ( trainval, deploy, and solver) ? Thanks a lot!

**Comment by dopplesoldner** [comment]
@gjy3035 There was a commit 3 days ago with some sample prototxts for train, test and solver. Check under models/pvanet directory

**Comment by gjy3035** [comment]
@dopplesoldner Thanks your reminder! But how do I select the proper super parameter? such as learning rate, momentum and so on?

**Comment by dopplesoldner** [comment]
@gjy3035 The hyperparameters are usually optimised based on the performance on a validation set. I'd start with the ones in the solver.protoxt and finetune later once you manage to train a model end-to-end.

**Comment by gjy3035** [comment]
@dopplesoldner In the model/pvanet/imagenet, I do not find the solver.protoxt. can you provide yours? Thanks a lot!

**Comment by dopplesoldner** [comment]
The sample prototxts can be found under https://github.com/sanghoon/pva-faster-rcnn/tree/master/models/pvanet/example_train_384

**Comment by gjy3035** [comment]
@dopplesoldner Thanks!

**Comment by sanghoon** [comment]
Hi, @dopplesoldner 

Did you set 'TRAIN.SCALE_MULTIPLE_OF' and 'TEST.SCALE_MULTIPLE_OF'?
They should be 32 because of hyper-feature concatenation.
Please refer to models/pvanet/cfgs/train.yml

**Comment by dopplesoldner** [comment]
@sanghoon Thanks, I managed to figure this out after using the cfgs in your last commit. Thanks for sharing

**Comment by hengck23** [comment]
@neodooth 
Hi, I am trying to use mxnet (based on https://github.com/precedenceguo/mx-rcnn) to repeat results of pvanet. Mxnet support multi-gpu training. 
Do you have good results on mxnet based pvanet?
